Feature: inactivity timeout and tests

divyenduz opened the pull request:
<!--
  Thanks for filing a pull request!

  Please look at the following checklist to ensure that your PR
  can be accepted quickly:
-->

TODO:

- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change

<!--**Pull Request Labels**

While not necessary, you can help organize our pull requests by labeling this issue when you open it.  To add a label automatically, simply [x] mark the appropriate box below:

- [ ] feature
- [ ] blocking
- [ ] docs

To add a label not listed above, simply place `/label another-label-name` on a line by itself.
-->

divyenduz commented:
Fixes https://github.com/apollographql/subscriptions-transport-ws/issues/351

timsuchanek commented:
Yeah! This is really cool! Looks good to me!
Is this good to merge @NeoPhi ?

NeoPhi commented on src/client.ts:
```diff
@@ -167,6 +174,9 @@ export class SubscriptionClient {
 
     let opId: string;
 
+    this.activeSubscriptions += 1;
```
This should be next to where `this.operations[opId] =` is done to ensure the reference counting it done correctly. Or possibly the check should just be against `Object.keys(this.operations).length` to avoid needing to do reference counting.

divyenduz replied:
Done, we are not counting `activeSubscriptions` anymore and relying on `Object.keys(this.operations).length` to test inactivity.

NeoPhi commented on src/client.ts:
```diff
@@ -167,6 +174,9 @@ export class SubscriptionClient {
     let opId: string;
 
+    this.activeSubscriptions += 1;
+    this.clearInactivityTimeout();
```
If someone is using `lazy = false` it isn't clear to me that this would then get triggered since a request never comes in.

NeoPhi commented on src/client.ts:
```diff
@@ -352,6 +362,23 @@ export class SubscriptionClient {
+    }
+  }
+
+  private setInactivityTimeout() {
```
What should the behavior of this function be as it related to `lazy = false` since `unsubscribe` is the only place it gets called.

divyenduz replied:
Should this function also test for connection to exist? I think this function is idempotent as is and can be called in any state. Not sure on the impact of `lazy` flag on this? Can you please elaborate.

NeoPhi commented on src/client.ts:
```diff
@@ -352,6 +362,23 @@ export class SubscriptionClient {
+    if (this.inactivityTimeout > 0 && this.activeSubscriptions === 0) {
+      this.inactivityTimeoutId = setTimeout(() => {
+        if (this.activeSubscriptions === 0) {
+          this.close();
```
Should add a test to verify that if the connection is closed due to inactivity, a future subscribe will correctly reopen the connection.

divyenduz replied:
Thanks, added the relevant test case.

NeoPhi commented on src/client.ts:
```diff
@@ -352,6 +362,23 @@ export class SubscriptionClient {
     }
   }
 
+  private clearInactivityTimeout() {
```
Should there be a call to this if the connection is closed for any other reason, otherwise I could see the reconnect logic fighting with the inactivity timeout and possibly tripping the flags for `closedByUser`.

divyenduz replied:
I agree, if a connection is closed (say, manually) but inactivity timeout had started. It might close the connection after `reconnection`.

To get around this, I called `clearInactivityTimeout` inside the `close` function.

divyenduz commented:
@NeoPhi : Thanks for the feedback

1. Updated the code to rely on `Object.keys(this.operations).length` instead of manually counting the references.

2. Called `clearInactivityTimeout` on `close` to avoid it messing up with reconnect logic. 

3. Added the test case to check reconnection after connection was closed due to inactivity. 

Please let me know if any further changes are needed.

timsuchanek commented:
@NeoPhi any news on this? Sorry for bothering, but we really need this to be merged as all people using server-side subscriptions with Yoga have connection leaks right now.

NeoPhi commented:
@timsuchanek I would contend that if a well behaved client is causing memory leaks in Yoga, that sounds like a server side issue. Long lived idle WebSockets are fine.
